### DISCUSSION TITLE
12-hour time format (Canadian localization)

### DISCUSSION THREAD

**Discussion by candideu** [pretix/pretix#3842] [Ideas]
I'd like the option to display time in a 12-hour time format (ex: 3:00pm)

**Comment by raphaelm** [comment]
just set the "region" of your event to US :)

**Reply by candideu** [reply]
I'd like this to be set for Canada as well. We also use this format, and my event is in Canada.

**Reply by pc-coholic** [reply]
Seems to me, like [the Canadian Government recommends the 24 hour notation while only discouraging the 12h notation](https://en.m.wikipedia.org/w/index.php?title=Date_and_time_notation_in_Canada&oldformat=true#Time) - also matches my personal experience from living in Canada, that the 12 hour system is only used in colloquial usage (and TV announcements), whereas 24 hour syntax is used in official communication.

So for now, selecting Canada as the region will leave things "as is".

But even if your event is in Canada and you set the region to US, nothing bad will happen: as far as I can tell, you'll get exactly what you want: 12 hour display and north-american formatting for numbers, monetary amounts, etc...

**Reply by candideu** [reply]
Hi @pc-coholic, 

The Wikipedia article you linked to cites [resources that have been marked by the government as outdated and archived](https://www.btb.termiumplus.gc.ca/tcdnstyl-chap?lang=eng&lettr=chapsect5&info0=5.13). The [up-to-date language tips shared by the government](https://www.noslangues-ourlanguages.gc.ca/en/writing-tips-plus/time-of-day-elapsed-time.html) have guidance for both formats. It's really only relevant, as you've said, for official government communication — and even those don't necessarily follow the 24hr time format:

### [House of Commons](https://www.ourcommons.ca/procedure/our-procedure/TypicalSittingDay/c_g_typicalsittingday-e.html) (and their [official House publications](https://www.ourcommons.ca/documentviewer/en/44-1/house/sitting-269A/hansard)):
![https://www.ourcommons.ca/procedure/our-procedure/TypicalSittingDay/c_g_typicalsittingday-e.html](https://github.com/pretix/pretix/assets/55474996/659b861b-d92a-4fc0-8d6d-2712cba5b79b)

### [Canada Revenue Agency](https://www.canada.ca/en/revenue-agency/services/e-services/digital-services-individuals/netfile-overview/hours-service.html):
![https://www.canada.ca/en/revenue-agency/services/e-services/digital-services-individuals/netfile-overview/hours-service.html](https://github.com/pretix/pretix/assets/55474996/e6d8eab9-f96b-4506-809a-e944b3c499a8)

### [National Defence (Military)](https://www.canada.ca/en/department-national-defence/services/benefits-military/military-housing/contact-cfha.html)
![image](https://github.com/pretix/pretix/assets/55474996/1e29c4f7-57fc-46bc-a636-ddc4413ec5d8)

### [Senate of Canada](https://sencanada.ca/en/calendar/) (and [the official reports for Senate debates](https://sencanada.ca/content/sen/chamber/441/debates/pdf/170db_2023-12-12-e.pdf)):
![image](https://github.com/pretix/pretix/assets/55474996/a49dbb11-4651-4179-b5e3-0bf017b83dd2)

---

I was born and raised in Canada, have lived across the country, have worked in government, and speak both French and English fluently. The use of 24h time is really only used by Francophones, who [represent 22.8% of the population](https://www.canada.ca/en/canadian-heritage/services/official-languages-bilingualism/publications/facts-canadian-francophonie.html), and the majority live in the province of Québec. Outside of that, 12hr is more widespread.

I suggest:
- defaulting to 12hr,
- offering both options if "Canada" is selected,
- having a Canadian English region/locale and Canadian French region/locale, or
- rather than having a general "Region" field in the "Localization" tab (which makes it difficult for users to know exactly what formats are being used): why not allow each organizer to select the date display, currency, numbering, and time formats of their choice as separate fields, like Pretix does for customer names?

---

> But even if your event is in Canada and you set the region to US, nothing bad will happen: as far as I can tell, you'll get exactly what you want: 12 hour display and north-american formatting for numbers, monetary amounts, etc...

There should be no reason to have to use a different region if "Canada" is on the list.

![image](https://github.com/pretix/pretix/assets/55474996/7ab7602a-f508-496e-9308-a637864a7117)

Additionally, the "Region" field indicates that it's used to set:
* date formatting -- which differs between Canada (day/month/year) and US (month/day/year), and 
* "default country for customer addresses" -- which is inconvenient for an in-person event hosted in Canada with attendees who primarily reside in Canada.

I tried to find more documentation for "Region", but [didn't have much luck](https://docs.pretix.eu/en/latest/search.html?q=region&check_keywords=yes&area=default). If this field is also being used for numbers and monetary amounts, as you've stated, that information should also be included in the field's help text.

---

> north-american formatting for numbers, monetary amounts, etc...

The formatting for numbers and monetary amounts is different for [Canadian French](https://www.noslangues-ourlanguages.gc.ca/fr/cles-de-la-redaction/mille-million-milliard.html) vs [Canadian English](https://www.noslangues-ourlanguages.gc.ca/en/writing-tips-plus/numbers-decimal-fractions.html) ("400 000 $" vs "$400,000").

That said, that might have to be a larger discussion on Canadian localization, along with Canadian English spelling (which differs from UK and US English), formal vs informal French, using "tu" vs "vous", etc... For now, I just want a 12hr time format. 😅

**Reply by pc-coholic** [reply]
This seems like it would end up in a larger discussion what is right and what is wrong, including regional and personal preferences :grimacing: 

For the most impact, feel free to file a PR against the django core like it has been done here: https://github.com/django/django/pull/1643

For a more immediate solution that only applies to pretix, you could try patching a en_CA into `pretix.helpers.formats.`, as has been done for `pretix.helpers.formats.en_US` in your local pretix: https://github.com/pretix/pretix/tree/master/src/pretix/helpers/formats (Just make sure to have provisions for when you update the pretix core...)

But in the bigger picture, I think there needs to be some consent first, what format is the one true one to rule them all... Not to be petty or anything: But I am not sure I have the stamina and/or professional knowledge to give the "win" to anyone at this pint ;-)

**Reply by candideu** [reply]
> For the most impact, feel free to file a PR against the django core like it has been done here: https://github.com/django/django/pull/1643

~~Will do.~~

Done: https://github.com/django/django/pull/17794

> For a more immediate solution that only applies to pretix, you could try patching a en_CA into pretix.helpers.formats., as has been done for pretix.helpers.formats.en_US in your local pretix: https://github.com/pretix/pretix/tree/master/src/pretix/helpers/formats (Just make sure to have provisions for when you update the pretix core...)

~~I'm happy to add a patch, but I'm not sure what is meant by "make sure to have provisions" as I'm not familiar with Django. Are there any other files that should be added/edited in pretix core?~~

See PR: https://github.com/pretix/pretix/pull/3844

> But in the bigger picture, I think there needs to be some consent first, what format is the one true one to rule them all... Not to be petty or anything: But I am not sure I have the stamina and/or professional knowledge to give the "win" to anyone at this pint ;-)

I don't think it's a matter of winning or losing: just looking for more clarity on what certain fields affect (in this case, "Region"), and giving users options.
